### ISSUE TITLE
batching support for ncon

### ISSUE THREAD

**Issue by mganahl** [google/TensorNetwork#640]
batched contractions are important for several machine learning like applications. `ncon` should support this as well

**Comment by Thenerdstation** [comment]
I think this will depend on #401 work.

**Comment by mganahl** [comment]
@gevenbly @Thenerdstation @alewis I have generalized `ncon` to support batched contractions, and I'd like to get your input regarding the interface. Currently `ncon` accepts a list of a list of labels (either `int` or `str`) for the argument `network_structure`. Any label prepended with a minus sign or a hyphen '-' is considered an output label, all other labels are contracted labels. `ncon` currently throws errors if any output label appears more than once, or any contracted label appears not exactly twice. To support batched contractions we have to relax these constraints. 

In particular, I'm considering to add a `batch_labels` argument. `batch_labels` is a list of labels passed by the user to flag labels as batch labels. The sole purpose of this argument is error checking by forcing the user to explicitly confirm that the implicit labeling in `network_structure` is correct. If `check_network=True`, the following additional checks are performed:

1. check if all elements in `batch_labels` appear in `network_structure`
1. check if all elements appearing more than twice in `network_structure` are flagged as `batch_labels`
3. check if all positive (or unhyphened) elements appearing exactly once in `network_structure` are flagged as `batch_labels`
4. check if all negative (or hyphened) elements appearing exactly twice in `network_structure` are flagged as `batch_labels`

If `check_network=False`, contracted labels and output labels are deduced entirely from the number of their appearances in `network_structure` together with their sign or hyphen, and `batch_labels` is ignored in this case. The user takes full responsibility for passing a correct `network_structure`.

For myself, I am not a big fan of enforcing the use of `batch_labels` and would prefer an `einsum` like notation by default, where everything is deduced entirely using minus signs/hyphens and number of appearances in `network_structure`, though I know that some people disagree. 

This means that `ncon` does almost the same thing as `einsum` except that defining the contraction order is much easier,
and we have more freedom in choosing label names (it also looks as if `ncon` is in many cases considerably faster than e.g. `np.einsum`).

Once this is supported we should also quickly add support for finding contraction paths for batched contractions.

Any input would be greatly appreciated.

**Comment by gevenbly** [comment]
@mganahl Could you please provide an example (or use case) of a batched contraction? Otherwise I am not precisely sure what is meant by this.

**Comment by mganahl** [comment]
Here is an example:
```python 
a = np.random.rand(1000,5,6)
b = np.random.rand(1000, 6,7)
res = np.matmul(a,b) #perform 1000 matrix-matrix muls 
print(res.shape) #prints (1000, 6, 7)
#einsum equivalent
np.einsum('kab,kbc->kac',a,b)
```
This one is more complicated:
```python 
a = np.random.rand(100,100,1000)
b = np.random.rand(100,1000, 100)
c = np.random.rand(100,1000,100)
np.einsum('abk,bkc,ckd->ad',a,b,c)
```
In machine learning settings such contractions appear e.g. when summing over training samples.

**Comment by gevenbly** [comment]
@mganahl Thanks for the examples. I would be more in favor of the approach of having the batch indices defined implicitly as -ve indices that appear more than once, rather than through a separate `batch_indices` input (although both ways of doing things have their own advantages).

**Comment by alewis** [comment]
I agree with Glen, although I know I was arguing the reverse a while back. On reflection it seems weird and clunky having an extra argument that literally has no effect except error checking, especially since einsum already sets a precedent.

**Comment by alewis** [comment]
The motivation for restricting Einstein notation to allow only pairs of contracted indices is so that one can easily read off whether an expression will transform as a tensor with respect to coordinate transformations when a metric is available. Since we obviously do not care about this there is no reason for the restriction.

**Comment by alewis** [comment]
I also think treating batching as a weird edge case given how ubiquitous it is in machine learning sends the wrong message about what sort of users we care about. Anyway it seems we all agree.

**Comment by Thenerdstation** [comment]
Agree with Adam and Glen. I think just having multiple uses of the '-x' will that that is required.

**Comment by mganahl** [comment]
great, in this case I'll make `batch_labels` an optional argument to enable better error checking

**Comment by mganahl** [comment]
be aware though of the following neccessary changes:
1. `ncon` will no longer throw an error if a negative label appears more than once
2. `ncon` will no longer throw an error if a positive label appears not exactly twice

**Comment by mganahl** [comment]
In that case it doesn't make sense to have `bach_labels` as an argument at all

**Comment by Thenerdstation** [comment]
SGTM

**Comment by gevenbly** [comment]
@mganahl I am curious how the new `ncon` handles the batched contractions internally? You say that it is faster than `einsum`? Is the ncon just doing a loop over all batch instances or something smarter than this?

**Comment by mganahl** [comment]
@gevenbly `ncon` is using `matmul` and `sum` operations. `matmul` exists in numpy,  jax and tensorflow where it can do batched matrix multiplications (for pytorch, we're using `torch.einsum` to emulate `matmul` behaviour).
It's not quite clear to me what the different versions of `einsum` are doing internally, but at least for numpy there are performance differences between using `np.matmul` and `np.sum` vs `np.einsum`. In some cases (e.g. small matrix dimensions, small batch dimensions, low-rank tensors) `np.einsum` seems to be faster than `np.matmul` + `np.sum`, but once bond dimensions/batch dimensions become large and/or tensors have higher rank, it looks like `ncon` is quite a bit faster.

**Comment by gevenbly** [comment]
Thanks for the explanation. In my previous ML codes I was using a `kron` based method to deal with batch indices (as to avoid having to loop over the batch indices explicitly), but it sees like python `matmul` is smart enough to do these batch contractions efficiently on its own.
It should be fairly easy to generalized the path solver to allow for batch indices (probably by using a second adjacency matrix that encodes the batch indices).

**Comment by mganahl** [comment]
@gevenbly can you quickly tell me how you did it with kron? It might be useful to benchmark the two ways

**Comment by gevenbly** [comment]
@mganahl Sure. A typical contraction that I would do for a ML problem is (where `d` is the samples in the batch and `chi` is a tn bond dim):
```
d = 10000
chi = 100
A = np.random.rand(d,chi)
B = np.random.rand(d,chi)
C = np.random.rand(chi,chi,chi)
T = np.einsum('ij,ik,jkl->il', A, B, C)
```
My solution (in MATLAB) to do this was using `kron` to avoid having an `for` loops, but an equivalent way to do the same thing is with numpy broadcasting and element-wise multiplications:
```
D = A.reshape(d,chi,1)*B.reshape(d,1,chi)
T = np.matmul(D.reshape(d,chi**2),C.reshape(chi**2,chi))
```
In my own benchmarks, the broadcasting method takes the same time as using `matmul` directly (which is why I think `matmul` must do something smarter than simply loop over the batch index, which would otherwise be slow).
